lib/backend/dir/impl.go

```diff
@@ -82,318 +82,306 @@ func New(params backend.Params) (backend.Backend, error) {
+// with the given TTL.
 func (bk *Backend) UpsertVal(bucket []string, key string, val []byte, ttl time.Duration) error {
-	// create the directory:
-	dirPath := path.Join(bk.RootDir, path.Join(bucket...))
-	err := os.MkdirAll(dirPath, defaultDirMode)
+	// Open the bucket to work on the items.
+	b, err := openBucket(bk.flatten(bucket), os.O_CREATE|os.O_RDWR)
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	filename := path.Join(dirPath, key)
-	f, err := os.OpenFile(filename, os.O_WRONLY|os.O_CREATE, defaultFileMode)
+	defer b.Close()
+
+	// Update the item in the bucket.
+	b.items[key] = bk.itemWithTTL(val, ttl)
+	b.itemsUpdated = true
+
+	return nil
+}
+
+// UpsertItems inserts (or updates if it already exists) all passed in
+// backend.Items with the given TTL.
+func (bk *Backend) UpsertItems(bucket []string, newItems []backend.Item) error {
+	// Open the bucket to work on the items.
+	b, err := openBucket(bk.flatten(bucket), os.O_CREATE|os.O_RDWR)
 	if err != nil {
-		if os.IsExist(err) {
-			return trace.AlreadyExists("%s/%s already exists", dirPath, key)
-		}
-		return trace.ConvertSystemError(err)
-	}
-	defer f.Close()
-	if err := utils.FSWriteLock(f); err != nil {
 		return trace.Wrap(err)
 	}
-	defer utils.FSUnlock(f)
-	if err := f.Truncate(0); err != nil {
-		return trace.ConvertSystemError(err)
-	}
-	n, err := f.Write(val)
-	if err == nil && n < len(val) {
-		return trace.Wrap(io.ErrShortWrite)
+	defer b.Close()
+
+	// Update items in bucket.
+	for _, e := range newItems {
+		b.items[e.Key] = bk.itemWithTTL(e.Value, e.TTL)
 	}
-	return trace.Wrap(bk.applyTTL(dirPath, key, ttl))
+	b.itemsUpdated = true
+
+	return nil
 }
 
 // GetVal return a value for a given key in the bucket
 func (bk *Backend) GetVal(bucket []string, key string) ([]byte, error) {
-	dirPath := path.Join(path.Join(bk.RootDir, path.Join(bucket...)))
-	filename := path.Join(dirPath, key)
-	expired, err := bk.checkTTL(dirPath, key)
+	// Open the bucket to work on the items.
+	b, err := openBucket(bk.flatten(bucket), os.O_RDWR)
 	if err != nil {
-		return nil, trace.Wrap(err)
-	}
-	if expired {
-		bk.DeleteKey(bucket, key)
-		return nil, trace.NotFound("key %q is not found", key)
-	}
-	f, err := os.OpenFile(filename, os.O_RDONLY, defaultFileMode)
-	if err != nil {
-		// GetVal() on a bucket must return 'BadParameter' error:
-		if fi, _ := os.Stat(filename); fi != nil && fi.IsDir() {
-			return nil, trace.BadParameter("%q is not a valid key", key)
+		// GetVal on a bucket needs to return trace.BadParameter. If opening the
+		// bucket failed a partial match up to a bucket may still exist. To support
+		// returning trace.BadParameter in this situation, loop over all keys in the
+		// backend and see if any match the prefix. If any match the prefix return
+		// trace.BadParameter, otherwise return the original error. This is
+		// consistent with our DynamoDB implementation.
+		files, er := ioutil.ReadDir(path.Join(bk.rootDir))
```

Comment:
we should not do a full scan of the bucket in this case, why not try to open the files with partial prefix instead?

Reply:
This is because `GetVal` has to return `trace.BadParameter` when specifically requesting a bucket. Take for example a backend with a single item in it:

```
/foo/bar/baz
```

If the bucket requested was "/foo/bar/baz" then we can just check for that particular bucket and return `trace.BadParameter` if we are able to open a bucket.

However, if the caller asks for "/foo" there is no way to differentiate the bucket does not exist at all from the caller requested a bucket that does exist (when we would return `trace.BadParameter`.